### PR TITLE
[Datasets] Add logical operator for randomize_block_order()

### PR DESCRIPTION
**Pull request by c21**
Signed-off-by: Cheng Su <scnju13@gmail.com>

## Why are these changes needed?

This PR adds logical operator for `randomize_block_order()`. The change includes:
1. Introduce `AbstractAllToAll` for all logical operators converted to `AllToAllOperator`
2. `RandomizeBlocks` logical operator for `randomize_block_order()`.
3. `_internal/planner` to move logic for `Planner` here and have generated function for `randomize_blocks`. This can be used later to create `MapOperator`/`AllToAllOperator`.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/logical/operators/map_operator.py [hunk 1]
```diff
@@ -62,6 +45,24 @@ def __init__(
         fn_constructor_kwargs: Optional[Dict[str, Any]] = None,
         ray_remote_args: Optional[Dict[str, Any]] = None,
     ):
+        """
```
Comment (c21) [inline python/ray/data/_internal/logical/operators/map_operator.py:48]: this is just moving around, no actual change for words.

File: python/ray/data/_internal/logical/operators/all_to_all_operator.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        self._supports_block_udf = supports_block_udf
+        self._block_udf = block_udf
+        self._ray_remote_args = ray_remote_args or {}
+
+
+def plan_all_to_all_op(
+    op: AbstractAllToAll,
+    input_physical_dag: PhysicalOperator
+) -> AllToAllOperator:
+    """Get the corresponding physical operators DAG for AbstractAllToAll operators."""
```
Comment (c21) [inline python/ray/data/_internal/logical/operators/all_to_all_operator.py:53]: This method is mostly copied from `legacy_compat.py:_stage_to_operator()`

File: python/ray/data/_internal/logical/operators/all_to_all_operator.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    """
+
+    def __init__(
+        self,
+        name: str,
+        input_op: LogicalOperator,
+        fn: Callable[[BlockList, bool, Callable, dict], Tuple[BlockList, dict]],
+        num_blocks: Optional[int] = None,
+        supports_block_udf: bool = False,
+        block_udf: Optional[BlockTransform] = None,
```
Comment (ericl) [inline python/ray/data/_internal/logical/operators/all_to_all_operator.py:24]: Should we remove supports_block_udf / block_udf from the logical plan (afaik it's only used to represent fused stages, and we can instead represent it as Map->AllToAll)?

I think the planner can generate the code implementing the block udf fusion.

File: python/ray/data/_internal/logical/operators/all_to_all_operator.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+class AbstractAllToAll(LogicalOperator):
+    """Abstract class for logical operators should be converted to physical
+    AllToAllOperator.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        input_op: LogicalOperator,
+        fn: Callable[[BlockList, bool, Callable, dict], Tuple[BlockList, dict]],
```
Comment (ericl) [inline python/ray/data/_internal/logical/operators/all_to_all_operator.py:21]: We shouldn't reference the legacy BlockList abstractions in the new packages.

File: python/ray/data/_internal/logical/operators/all_to_all_operator.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+from ray.data._internal.block_list import BlockList
+from ray.data._internal.compute import BlockTransform
+from ray.data._internal.execution.interfaces import PhysicalOperator, RefBundle
+from ray.data._internal.execution.operators.all_to_all_operator import AllToAllOperator
+from ray.data._internal.logical.interfaces import LogicalOperator
+
+from ray.data._internal.stats import StatsDict
+
+
+class AbstractAllToAll(LogicalOperator):
```
Comment (ericl) [inline python/ray/data/_internal/logical/operators/all_to_all_operator.py:12]: Should logical operators not specify a `fn` at all? That should be generated during planning time.

Comment (c21) [reply python/ray/data/_internal/logical/operators/all_to_all_operator.py:11]: @ericl - yeah agreed. Created a subdirectory `execution/operators/functions` to put the generated functions for each operator.

File: python/ray/data/_internal/logical/operators/all_to_all_operator.py (hunk 4 above)
Comment (c21) [reply python/ray/data/_internal/logical/operators/all_to_all_operator.py]: @ericl - removed.

File: python/ray/data/_internal/logical/operators/all_to_all_operator.py (hunk 3 above)
Comment (c21) [reply python/ray/data/_internal/logical/operators/all_to_all_operator.py]: @ericl - yeah removed.

File: python/ray/data/_internal/logical/operators/all_to_all_operator.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        super().__init__(name, [input_op])
+        self._fn = fn
+        self._num_blocks = num_blocks
+        self._supports_block_udf = supports_block_udf
+        self._block_udf = block_udf
+        self._ray_remote_args = ray_remote_args or {}
+
+
+def plan_all_to_all_op(
+    op: AbstractAllToAll, input_physical_dag: PhysicalOperator
```
[the remark this answers is not in the document]
Comment (c21) [reply python/ray/data/_internal/logical/operators/all_to_all_operator.py]: @jianoaix - sure, updated.

File: python/ray/data/_internal/logical/operators/map_operator.py [hunk 7]
```diff
@@ -62,6 +45,24 @@ def __init__(
...
+        Args:
+            name: Name for this operator. This is the name that will appear when
+                inspecting the logical plan of a Dataset.
+            input_op: The operator preceding this operator in the plan DAG. The outputs
+                of `input_op` will be the inputs to this operator.
+            block_fn: The transform function to apply to each input block to produce
```
Comment (jianoaix) [inline python/ray/data/_internal/logical/operators/map_operator.py:54]: Shall we name this as `transform_fn`? It's consistent (and IMO more clear) with the naming in new backend.

Comment (c21) [reply python/ray/data/_internal/logical/operators/map_operator.py:47]: yeah, actually after second thought, this argument should not be part of logical operator. Logical operator should only care about user-defined function. This is internal function we generate for execution. Plan to remove it and put into `planner` directory in a separate PR.

File: python/ray/data/_internal/planner/plan_map_op.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+import ray
+from ray.data._internal.compute import ActorPoolStrategy, TaskPoolStrategy, get_compute
+from ray.data._internal.execution.interfaces import PhysicalOperator
+from ray.data._internal.execution.operators.map_operator import MapOperator
+from ray.data._internal.logical.operators.map_operator import AbstractMap
+from ray.data.block import Block, CallableClass
+
+
+def _plan_map_op(op: AbstractMap, input_physical_dag: PhysicalOperator) -> MapOperator:
```
Comment (c21) [inline python/ray/data/_internal/planner/plan_map_op.py:11]: This is moved from `map_operator.py`.

File: python/ray/data/_internal/planner/plan_read_op.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import ray.cloudpickle as cloudpickle
+from ray.data._internal.execution.interfaces import PhysicalOperator, RefBundle
+from ray.data._internal.execution.operators.map_operator import MapOperator
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+from ray.data._internal.logical.operators.read_operator import Read
+from ray.data.block import Block, BlockMetadata
+from ray.data.datasource.datasource import ReadTask
+
+
+def _plan_read_op(op: Read) -> PhysicalOperator:
```
Comment (c21) [inline python/ray/data/_internal/planner/plan_read_op.py:13]: This is moved from `read_operator.py`

File: python/ray/data/_internal/planner/randomize_blocks.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+from ray.data._internal.execution.interfaces import RefBundle
+from ray.data._internal.stats import StatsDict
+
+
+def generate_randomize_blocks_fn(
+    seed: Optional[int],
+) -> Callable[[List[RefBundle]], Tuple[List[RefBundle], StatsDict]]:
+    """Generate function to randomize order of blocks."""
+
+    def fn(refs: List[RefBundle]) -> Tuple[List[RefBundle], StatsDict]:
```
Comment (c21) [inline python/ray/data/_internal/planner/randomize_blocks.py:12]: This function is refactored from
https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/stage_impl.py#L80 and 
https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/block_list.py#L228

File: python/ray/data/_internal/planner/plan_map_op.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    """Get the corresponding physical operators DAG for AbstractMap operators.
+
+    Note this method only converts the given `op`, but not its input dependencies.
+    See Planner.plan() for more details.
+    """
+    compute = get_compute(op._compute)
+    block_fn = op._block_fn
```
Comment (ericl) [inline python/ray/data/_internal/planner/plan_map_op.py:18]: Should we also remove block_fn from the op?

**Review by ericl — APPROVED** [review]

File: python/ray/data/_internal/planner/plan_map_op.py (hunk 11 above)
Comment (c21) [reply python/ray/data/_internal/planner/plan_map_op.py:18]: @ericl - yes, plan to do in a separate PR - https://github.com/ray-project/ray/pull/31977#discussion_r1089348501 .

**Review by jianoaix — APPROVED** [review]

**Pull request merged**